core/vm/vmcontext/system_evm.go

```diff
@@ -0,0 +1,120 @@
+package vmcontext
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/core/state"
+	"github.com/celo-org/celo-blockchain/core/types"
+	"github.com/celo-org/celo-blockchain/core/vm"
+)
+
+// ChainContext supports retrieving chain data and consensus parameters
+// from the blockchain to be used during transaction processing.
+type ExtendedChainContext interface {
+	vm.ChainContext
+
+	// GetVMConfig returns the node's vm configuration
+	GetVMConfig() *vm.Config
+
+	CurrentHeader() *types.Header
+
+	State() (*state.StateDB, error)
+}
+
```

Comment:
I think the below is pretty confusing with the proliferation of SystemEVM-related names (system, factory, and provider).  As I see it: SystemEVMFactory generates SystemEVM, which has a provider, which generates EVMs.  I'm not sure how to make the naming clearer.  If we can't, maybe we should include some comments to lay it out for readers.  It might also help to make `NewSystemEVM` and `NewSystemEVMForCurrentBlock` private in the same way that `newEVMProvider` is private, to delineate functions which are intended as part of the API from helper functions (I'm assuming those two are in fact just helper functions and not intended to be called directly elsewhere).  And maybe even place them separately in the code.  Granted, none of this would be a huge boost to the readability, but perhaps still recommendable in the absence of a better way to simplify the things in this file.